src/Nethermind/Nethermind.Serialization.Rlp/TxDecoder.cs

```diff
@@ -115,22 +115,28 @@ protected virtual T NewTx()
                         break;
                 }
 
-                if ((rlpBehaviors & RlpBehaviors.AllowExtraBytes) != RlpBehaviors.AllowExtraBytes)
+                if ((rlpBehaviors & RlpBehaviors.AllowExtraBytes) == 0)
                 {
                     rlpStream.Check(positionAfterNetworkWrapper);
                 }
 
-                transaction.Hash = CalculateHashForNetworkPayloadForm(transaction.Type, transactionSequence);
-            }
-            else if (transactionSequence.Length <= TxDecoder.MaxDelayedHashTxnSize && _lazyHash)
-            {
-                // Delay hash generation, as may be filtered as having too low gas etc
-                transaction.SetPreHashNoLock(transactionSequence);
+                if ((rlpBehaviors & RlpBehaviors.ExcludeHashes) == 0)
+                {
+                    transaction.Hash = CalculateHashForNetworkPayloadForm(transaction.Type, transactionSequence);
+                }
             }
-            else
+            else if ((rlpBehaviors & RlpBehaviors.ExcludeHashes) == 0)
             {
-                // Just calculate the Hash as txn too large
-                transaction.Hash = Keccak.Compute(transactionSequence);
+                if (transactionSequence.Length <= TxDecoder.MaxDelayedHashTxnSize && _lazyHash)
+                {
+                    // Delay hash generation, as may be filtered as having too low gas etc
+                    transaction.SetPreHashNoLock(transactionSequence);
```

Comment:
I think we could change `transactionSequence` here to be `ReadOnlyMemory`, then we would't have to copy it later to mempool array? But then we would have to keep the RlpStream alive if it is based on donetty buffer?

Reply:
Probably not for this PR, but something to explore

Reply:
Another Rlp behaviour? 😉